telemetry/data_slow_query.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/pingcap/tidb/domain/infosync"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	pmodel "github.com/prometheus/common/model"
+	"go.uber.org/zap"
+)
+
+type slowQueryStats struct {
+	// Slow Query statistic buckets
+	SQBInfo *SlowQueryBucket `json:"slowQueryBucket"`
+}
+
+// SlowQueryBucket records the statistic information of slow query buckets
+// Buckets:   prometheus.ExponentialBuckets(0.001, 2, 28), // 1ms ~ 1.5days  // defined in metrics/server.go
+type SlowQueryBucket map[string]int
+
+const slowQueryBucketNum = 29 //prometheus.ExponentialBuckets(0.001, 2, 28), and 1 more +Inf
+
+var (
+	// LastSQBInfo records last statistic information of slow query buckets
+	LastSQBInfo SlowQueryBucket
+	// CurrentSQBInfo records current statitic information of slow query buckets
+	CurrentSQBInfo SlowQueryBucket
+)
+
+func getSlowQueryStats(ctx sessionctx.Context) (*slowQueryStats, error) {
+	slowQueryBucket, err := GetSlowQueryBucket(ctx)
+	if err != nil {
+		logutil.BgLogger().Info(err.Error())
+		return nil, err
+	}
+
+	return &slowQueryStats{slowQueryBucket}, nil
+}
+
+// GetSlowQueryBucket genenrates the delta SlowQueryBucket to report
+func GetSlowQueryBucket(ctx sessionctx.Context) (*SlowQueryBucket, error) {
+	// update CurrentSQBInfo first, then gen delta
+	if err := UpdateCurrentSQB(ctx); err != nil {
+		return nil, err
+	}
+	delta := CalculateDeltaSQB()
+	return delta, nil
+}
+
+// UpdateCurrentSQB records current slow query buckets
+func UpdateCurrentSQB(ctx sessionctx.Context) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = errors.New("unknown failure")
```

Comment:
err = errors.Trace(errors.Errorf())
```suggestion
			err = errors.Trace(errors.Errorf(fmt.Sprintln(r)))
```

Reply:
Trace is not declared by errors. 

Do you mean another package?

Thanks.

Reply:
pingcap/errors has this method 😄

Reply:
Many thanks, done.